### PR TITLE
Improve performance of discovering undeclared outputs

### LINKED ISSUE #29464: Improve performance of undeclared output discovery
### Description of the bug:

Generating the undeclared outputs manifest can become noticeably slow when tests generate a large number of undeclared outputs.

XCResult files from `xcodebuild` contain hundreds, or even thousands, of files within them for large test bundles and can take over 10 seconds to traverse with the current implantation.

Our project has a unit test bundle that takes roughly 12 seconds for the tests to complete, but then another 10-12 seconds is consumed by generating the undeclared output manifest for the XCResult file it creates.

### Which category does this issue belong to?

Performance

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Create a `sh_test` target that writes 2,000 files.

```bash
#!/usr/bin/env bash

set -euo pipefail

for i in $(seq 1 2000); do
  printf 'content %d\n' "$i" > "${TEST_UNDECLARED_OUTPUTS_DIR}/output_${i}.txt"
done

echo "Writing files finished"
```

Run the test with `--test_output=streamed` and observe the delay between "Writing files finished" being printed to the console and the test finishing.

### Which operating system are you running Bazel on?

macOS 26.4.1

### What is the output of `bazel info release`?

release 9.1.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?

```text
https://github.com/bazelbuild/bazel.git
42b1f15e1063cffdf9711394bbf8020c0d72e24a
```

### If this is a regression, please try to identify the Bazel commit where the bug was introduced with bazelisk --bisect.

_No response_

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

### PR DESCRIPTION
**Pull request by john-flanagan**
### Description

Improve the performance of generating the undeclared test output manifest by converting repeated calls to `stat` and `file` into a single command using `paste` and `xargs` to generate the same output.

I believe the relevant behavior of `test-startup.sh` is covered by [test_wrapper_test.py](https://github.com/bazelbuild/bazel/blob/283fec5e2ab5eba6f619ef7bfc5fdf16ab56451c/src/test/py/bazel/test_wrapper_test.py#L557-L581)

### Motivation

Fixes https://github.com/bazelbuild/bazel/issues/29464

### Build API Changes

No

### Checklist

- [x] I have added tests for the new use cases (if any).
- [x] I have updated the documentation (if applicable).

### Release Notes

RELNOTES: None

### CODE REVIEW

**Comment by john-flanagan** [comment]
I isolated just the updated portion of the script and pointed it at one of our XCResult files that has a noticeable delay when generating the manifest.

The updated version of the script is ~100x faster on my machine than the current implementation.

```bash
$ hyperfine --warmup 1 "TEST_UNDECLARED_OUTPUTS_DIR=output TEST_UNDECLARED_OUTPUTS_MANIFEST=updated-manifest.txt ./updated-test-setup.sh" "TEST_UNDECLARED_OUTPUTS_DIR=output TEST_UNDECLARED_OUTPUTS_MANIFEST=old-manifest.txt ./old-test-setup.sh"
Benchmark 1: TEST_UNDECLARED_OUTPUTS_DIR=output TEST_UNDECLARED_OUTPUTS_MANIFEST=updated-manifest.txt ./updated-test-setup.sh
  Time (mean ± σ):      77.9 ms ±   8.5 ms    [User: 118.0 ms, System: 29.3 ms]
  Range (min … max):    68.3 ms … 100.9 ms    39 runs

Benchmark 2: TEST_UNDECLARED_OUTPUTS_DIR=output TEST_UNDECLARED_OUTPUTS_MANIFEST=old-manifest.txt ./old-test-setup.sh
  Time (mean ± σ):      8.309 s ±  0.032 s    [User: 2.307 s, System: 4.519 s]
  Range (min … max):    8.255 s …  8.360 s    10 runs

Summary
  TEST_UNDECLARED_OUTPUTS_DIR=output TEST_UNDECLARED_OUTPUTS_MANIFEST=updated-manifest.txt ./updated-test-setup.sh ran
  106.63 ± 11.70 times faster than TEST_UNDECLARED_OUTPUTS_DIR=output TEST_UNDECLARED_OUTPUTS_MANIFEST=old-manifest.txt ./old-test-setup.sh
```

Hardware: M4 Max MacBook Pro

**Comment by fmeum** [comment]
Note that the test you linked only runs on Windows (https://github.com/bazelbuild/bazel/blob/283fec5e2ab5eba6f619ef7bfc5fdf16ab56451c/src/test/py/bazel/BUILD#L211) while the logic you changed is only relevant on Unix.

**Comment by john-flanagan** [comment]
@fmeum Good callout. [bazel_test_test.sh](https://github.com/bazelbuild/bazel/blob/283fec5e2ab5eba6f619ef7bfc5fdf16ab56451c/src/test/shell/bazel/bazel_test_test.sh#L1013-L1084) should cover this on Unix, right?

**Review by fmeum — APPROVED** [review]

File: tools/test/test-setup.sh [hunk 1]
```diff
@@ -323,15 +323,23 @@ if [[ -n "$TEST_UNDECLARED_OUTPUTS_DIR" && -n "$TEST_UNDECLARED_OUTPUTS_MANIFEST
...
+    else
+      stat_size_fmt='-c%s'
+    fi
+
+    # Path length + "/" + 1 because 'cut' is 1 indexed
+    cut_prefix_len=$(( ${#TEST_UNDECLARED_OUTPUTS_DIR} + 2 ))
+
+    paste \
+      <(printf '%s\n' "$undeclared_outputs" | cut -c"${cut_prefix_len}"-) \
+      <(printf '%s\n' "$undeclared_outputs" | tr '\n' '\0' | xargs -0 stat "$stat_size_fmt" 2>/dev/null) \
```
Comment (meisterT) [inline tools/test/test-setup.sh]: we need to make sure that we catch if things go wrong, e.g. something like this

```
<(printf '%s\n' "$undeclared_outputs" | tr '\n' '\0' | xargs -0 -I{} sh -c "stat $stat_size_fmt '{}' 2>/dev/null || echo 'Could not stat {}'") \
```

Similarly below.

File: tools/test/test-setup.sh [hunk 2]
```diff
@@ -323,15 +323,23 @@ if [[ -n "$TEST_UNDECLARED_OUTPUTS_DIR" && -n "$TEST_UNDECLARED_OUTPUTS_MANIFEST
...
+      stat_size_fmt='-f%z'
+    else
+      stat_size_fmt='-c%s'
+    fi
+
+    # Path length + "/" + 1 because 'cut' is 1 indexed
+    cut_prefix_len=$(( ${#TEST_UNDECLARED_OUTPUTS_DIR} + 2 ))
+
+    paste \
+      <(printf '%s\n' "$undeclared_outputs" | cut -c"${cut_prefix_len}"-) \
```
Comment (meisterT) [inline tools/test/test-setup.sh]: I think we should rather use `sed` instead of `cut` here to avoid issues with multi-byte unicode strings

**Review by meisterT — COMMENTED** [review]
Thanks for looking into this. Some suggestions to make the more robust.

File: tools/test/test-setup.sh (hunk 1 above)
Comment (john-flanagan) [reply tools/test/test-setup.sh]: I tried this approach just now but got `xargs: command line cannot be assembled, too long`

The `find` command is falling back to outputting an error message using `awk`, but that's all or nothing.

We could do something similar for both if that approach is enough:
```bash
      <(printf '%s\n' "$undeclared_outputs" | tr '\n' '\0' | xargs -0 stat $stat_size_fmt \
          || awk '{print "Could not establish file size"}' <<< "$undeclared_outputs") \
      <(printf '%s\n' "$undeclared_outputs" | tr '\n' '\0' | xargs -0 file -L -b --mime-type \
          || awk '{print "Could not establish file type"}' <<< "$undeclared_outputs") \
```

Maybe a separate issue, but the existing error message for `find` errors is long and repetitive since it includes the full path to the undeclared output file. Since the first column of the manifest is already the output name it seems like we could reasonably drop the full file path from the error message.

Comment (john-flanagan) [reply tools/test/test-setup.sh]: I pushed up this approach. I'm happy to iterate if wanted.

File: tools/test/test-setup.sh (hunk 2 above)
Comment (john-flanagan) [reply tools/test/test-setup.sh]: Switching to `sed` cut the performance by ~50%. I tried `awk` as well, but it had the same impact.

It's still plenty fast though:
```bash
$ hyperfine -r 40 --warmup 1 "TEST_UNDECLARED_OUTPUTS_DIR=output TEST_UNDECLARED_OUTPUTS_MANIFEST=feedback-manifest.txt ./feedback-test-setup.sh"
Benchmark 1: TEST_UNDECLARED_OUTPUTS_DIR=output TEST_UNDECLARED_OUTPUTS_MANIFEST=feedback-manifest.txt ./feedback-test-setup.sh
  Time (mean ± σ):     174.9 ms ±   5.0 ms    [User: 157.8 ms, System: 79.9 ms]
  Range (min … max):   166.7 ms … 189.6 ms    40 runs
```

File: tools/test/test-setup.sh (hunk 1 above)
Comment (meisterT) [reply tools/test/test-setup.sh]: Thanks, I think it doesn't fully work in tricky edge case scenarios that the `while` loop handled naturally. I need to think more about how to properly fix it.

The issue is that `stat`  doesn't print anything to stdout when it fails on a file. If we use `xargs`  and one file in a batch fails the line count for the sizes column will shrink, causing `paste` to misalign the data and associate sizes with the wrong files.

`|| awk`  fallback unfortunately doesn't fix these partial success scenarios:
If `file1` succeeds and `file2` fails, `xargs` will emit the size of `file1` and then `awk` will emit error lines for **all** files. This results in too many lines in the size column, again causing `paste` to misalign the data.

We need another way to guarantee exactly one output line for every file even on failure.

**Comment by adincebic** [comment]
@john-flanagan @meisterT friendly ping. This would be nice to land.
